### PR TITLE
[SFN] Enhanced Support for Exception Handling in Parallel States

### PR DESCRIPTION
**Pull request by MEPalma**
## Motivation
The SFN v2 interpreter is currently unable to propagate Parallel states' Branch failures to the Parallel state. This masks state failures, and prevents exception handling. The PR resolves such issues, ensuring that Branch failures are relayed orderly to the state's thread.
Closes: #9863 

## Changes
Reworked the role of the Branches evaluation component to dispatch to a branch worker pool, which intercepts through callbacks workers' terminations. Workers' failures are cached and other workers are stopped. It then signals the state's thread to resume computation whenever a stopping condition is met (all branches evaluated, or at least one branch failed).
Reworked program workers into branch workers, which integrate better with branches objects.
Reworked the eval state of parallel states to orderly handle retry and/or catch declarations.
Added a few relevant tests.

### CODE REVIEW

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_parallel/branch_worker.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        LOG.info(f"[BranchWorker] [launched] [id: {self._worker_thread.native_id}]")
+        self._program.eval(self.env)
+        LOG.info(f"[BranchWorker] [terminated] [id: {self._worker_thread.native_id}]")
+        self._branch_worker_comm.on_terminated(env=self.env)
+
+    def start(self):
+        if self._worker_thread is not None:
+            raise RuntimeError(f"Attempted to rerun BranchWorker for program ${self._program}.")
+
+        self._worker_thread = threading.Thread(target=self._thread_routine)
```
Comment (dominikschubert) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_parallel/branch_worker.py:41]: Please make sure to always name your threads and thread pools. This practice helps us identify any potential leaks

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_parallel/branches_decl.py [hunk 2]
```diff
@@ -1,41 +1,116 @@
...
+from localstack.services.stepfunctions.asl.eval.program_state import ProgramError, ProgramState
+from localstack.utils.collections import select_from_typed_dict
+
+
+class BranchWorkerPool(BranchWorker.BranchWorkerComm):
+    _mutex: Final[threading.Lock]
+    _termination_event: Final[threading.Event]
+    _active_workers_num: int
+
+    _terminated: bool
```
Comment (dominikschubert) [inline localstack/services/stepfunctions/asl/component/state/state_execution/state_parallel/branches_decl.py:30]: isn't this equivalent to checking if the `_termination_event` is set?

**Review by dominikschubert — APPROVED** [review]
I'd like to give the whole concurrency setup a more detailed review in a month or so, but for now it's looking good and besides the missing naming it LGTM :+1: 

Good coverage with the new tests :+1:

File: localstack/services/stepfunctions/asl/component/state/state_execution/state_parallel/branches_decl.py (hunk 2 above)
Comment (MEPalma) [reply localstack/services/stepfunctions/asl/component/state/state_execution/state_parallel/branches_decl.py:30]: Indeed, thank you!

**Pull request merged**